### PR TITLE
New TF loading weights

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR improves the way we load the TensorFlow weights. Before we had to go through the instantiated model + the checkpoint twice:
- once for loading the weights from the checkpoints into the instantiated model
- once for computing the missing and unexpected keys

Now both are done simultaneously which makes the loading faster.

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -318,8 +307,15 @@ def load_tf_weights(model, resolved_archive_file):
 
                         weight_value_tuples.append((symbolic_weight, array))
 
+                    symbolic_weights_names.add(symbolic_weight_name)
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_utils.py:310]: This will add the `symbolic_weight_name` to `symbolic_weights_names` even if `saved_weight_value` is `None`. Is that expected?

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -253,60 +253,49 @@ def detect_tf_missing_unexpected_layers(model, resolved_archive_file):
...
+        saved_h5_model_layers_name = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
+        model_layers_name_value = {}
 
         for layer in model.layers:
-            if layer.name in saved_layer_names:
-                g = f[layer.name]
-                saved_weight_names = hdf5_format.load_attributes_from_hdf5_group(g, "weight_names")
-                saved_weight_names_set = set(
-                    "/".join(weight_name.split("/")[2:]) for weight_name in saved_weight_names
-                )
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_utils.py:267]: This line and the line defining `symbolic_weights_names` were removed. Instead of being filled by
```py
"/".join(weight_name.split("/")[2:]) for weight_name in saved_weight_names
```
they will now be filled by
```py
"/".join(weight_name.split("/")[1:]) for weight_name in saved_weight_names
```
Why the change from `2:` to `1:`?

**Review by LysandreJik — APPROVED** [review]
After some testing, this looks good to me. Given that it's a very dangerous area of the library I'll ask @sgugger or @patrickvonplaten to review as well.

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -236,15 +236,15 @@ def compute_loss(self, labels, logits):
         return loss_fn(next_sentence_label, next_sentence_reduced_logits)
 
 
-def detect_tf_missing_unexpected_layers(model, resolved_archive_file):
+def load_tf_weights(model, resolved_archive_file):
     """
-    Detect missing and unexpected layers.
+    Detect missing and unexpected layers
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:241]: The docstring should be adapted since this function loads the TF weights **and** detects missing and unexpected layers.

File: src/transformers/modeling_tf_utils.py [hunk 4]
```diff
@@ -236,15 +236,15 @@ def compute_loss(self, labels, logits):
...
 
     Args:
         model (:obj:`tf.keras.models.Model`):
             The model to load the weights into.
         resolved_archive_file (:obj:`str`):
-            The location of the H5 file.
+            The location of the H5 file
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:247]: No need to remove this dot.

File: src/transformers/modeling_tf_utils.py [hunk 5]
```diff
@@ -253,60 +253,49 @@ def detect_tf_missing_unexpected_layers(model, resolved_archive_file):
     unexpected_layers = []
 
     with h5py.File(resolved_archive_file, "r") as f:
-        saved_layer_names = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
-        model_layer_names = set(layer.name for layer in model.layers)
-        missing_layers = list(model_layer_names - saved_layer_names)
-        unexpected_layers = list(saved_layer_names - model_layer_names)
+        saved_h5_model_layers_name = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
+        model_layers_name_value = {}
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:257]: This line and the for loop after should be replaced by
```
model_layers_name_value = {layer.name: layer for layer in model.layers}
```
This is easier code to read IMO and will be faster (dict comprehension is faster than a manual loop).

File: src/transformers/modeling_tf_utils.py [hunk 6]
```diff
@@ -253,60 +253,49 @@ def detect_tf_missing_unexpected_layers(model, resolved_archive_file):
...
+        model_layers_name = set(model_layers_name_value.keys())
+        renamed_saved_h5_model_layers_names = set()
 
+        for layer_name in saved_h5_model_layers_name:
+            name = layer_name
 
-def load_tf_weights(model, resolved_archive_file):
-    """
-    Load the TF weights from a H5 file.
+            renamed_saved_h5_model_layers_names.add(name)
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:269]: I don't get this code. It loops through the set `saved_h5_model_layers_name` and add each element to a new empty set. So this is exactly the same as doing
```
renamed_saved_h5_model_layers_names = saved_h5_model_layers_name.copy()
```
but why do we need a new variable and why is it "renamed"?

File: src/transformers/modeling_tf_utils.py [hunk 7]
```diff
@@ -253,60 +253,49 @@ def detect_tf_missing_unexpected_layers(model, resolved_archive_file):
...
                 for symbolic_weight in symbolic_weights:
-                    splited_layers = symbolic_weight.name.split("/")[1:]
-                    symbolic_weight_name = "/".join(splited_layers)
+                    symbolic_weight_name = "/".join(symbolic_weight.name.split("/")[1:])
+                    saved_weight_value = None
 
                     if symbolic_weight_name in saved_weight_names_values:
                         saved_weight_value = saved_weight_names_values[symbolic_weight_name]
 
+                    if saved_weight_value is not None:
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:298]: Why adding this test and the line before `saved_weight_value = None`? There is no other place `saved_weight_value` is used so this just adds extra complexity in the code.

File: src/transformers/modeling_tf_utils.py [hunk 8]
```diff
@@ -253,60 +253,49 @@ def detect_tf_missing_unexpected_layers(model, resolved_archive_file):
...
+        missing_layers = list(model_layers_name - renamed_saved_h5_model_layers_names)
+        unexpected_layers = list(renamed_saved_h5_model_layers_names - model_layers_name)
+        saved_weight_names_set = set()
+        symbolic_weights_names = set()
         weight_value_tuples = []
 
-        for layer in model.layers:
-            if layer.name in saved_layer_names:
-                g = f[layer.name]
+        for layer_name in saved_h5_model_layers_name:
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:277]: Any reason to change the loop? AFAICT we only do things for objects that are in the intersection of `saved_h5_model_layer_names` (previously `saved_layer_names`) and the `model_layer_names`. However, the previous loop did not require the dictionary `model_layers_name_value` and this dictionary is only used here, so it seems we only gain complexity with this change.

**Review by sgugger — COMMENTED** [review]
While doing the two things in one loop instead of two is definitely something we should do, I don't understand a lot of extra-complexity added by this code:
- There is a new `renamed_saved_h5_model_layers_names` introduced and built by a for loop that ends up being a copy of `saved_h5_model_layers_name`.
- A new dictionary is introduced just because you loop through the `saved_h5_model_layers_name` instead of the `model.layers` like before.

The same goal can be achieved by staying closer to the original code and not adding that complexity. 

I also have the same question as @LysandreJik on the `2:` becoming `1:`, but it seems to be working locally.

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:267]: I took a deeper look and understand the change here. I guess it's because the second index is the the prefix, and since in TF the main layer is named after the prefix, it will remain the same across base models and models with heads.

**Comment by jplu** [comment]
I have added a lot of comments in the method to make it clearer. @LysandreJik @sgugger it should be easier to understand now.

File: src/transformers/modeling_tf_utils.py [hunk 9]
```diff
@@ -236,78 +236,90 @@ def compute_loss(self, labels, logits):
...
-                    "/".join(symbolic_weight.name.split("/")[2:]) for symbolic_weight in symbolic_weights
-                )
-                missing_layers.extend(list(symbolic_weights_names - saved_weight_names_set))
-                unexpected_layers.extend(list(saved_weight_names_set - symbolic_weights_names))
+        # Retrieve the name of each layer from the H5 file
+        saved_h5_model_layers_name = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
+        model_layers_name_value = {}
 
-    return missing_layers, unexpected_layers
+        # Retrieve the name of each layer from the instanciated model
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:261]:
```suggestion
        # Retrieve the name of each layer from the instantiated model
```

File: src/transformers/modeling_tf_utils.py [hunk 10]
```diff
@@ -236,78 +236,90 @@ def compute_loss(self, labels, logits):
...
+        # make it a dict that looks like {"layer_name": Layer object}
+        model_layers_name_value = {layer.name: layer for layer in model.layers}
 
+        # Create a set of unique names of the layers that come from the instanciated model
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:265]:
```suggestion
        # Create a set of unique names of the layers that come from the instantiated model
```

File: src/transformers/modeling_tf_utils.py (hunk 7 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_tf_utils.py:318]: Agree here I think we can just put all the code below `if saved_weight_value is not None:` directly under `if symbolic_weight_name in saved_weight_names_values:` no?

File: src/transformers/modeling_tf_utils.py [hunk 11]
```diff
@@ -236,78 +236,90 @@ def compute_loss(self, labels, logits):
...
+            if layer_name in model_layers_name:
+                # Get layer_name from the H5 file
+                g = f[layer_name]
+                # Get all the weights that are attach to layer_name
                 saved_weight_names = hdf5_format.load_attributes_from_hdf5_group(g, "weight_names")
+                # Get the layer object from the layer_name in the dict that represents the instanciated model
+                layer = model_layers_name_value[layer_name]
+                # Get all the weights as a list from the layer object
                 symbolic_weights = layer.trainable_weights + layer.non_trainable_weights
                 saved_weight_names_values = {}
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:290]: it's a nit, but I think it's cleaner to name this dict just `saved_weights`. `saved_weight_names_values` makes me think of a list of values not a key/value dict...

File: src/transformers/modeling_tf_utils.py [hunk 12]
```diff
@@ -236,78 +236,90 @@ def compute_loss(self, labels, logits):
...
-        for layer in model.layers:
-            if layer.name in saved_layer_names:
-                g = f[layer.name]
+        # Compute missing and unexpected sub layers
+        # Store the weights in list of tuples that looks like [(weight_object, value_of_weight),...]
+        for layer_name in saved_h5_model_layers_name:
+            # if layer_name from the H5 file belongs to the layers from the instanciated model
+            if layer_name in model_layers_name:
+                # Get layer_name from the H5 file
+                g = f[layer_name]
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:283]: what is `g`?

File: src/transformers/modeling_tf_utils.py (hunk 7 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_tf_utils.py:318]: Alternatively we could do `saved_weight_value = saved_weight_value.get(symbolic_weight_name, None)`

File: src/transformers/modeling_tf_utils.py (hunk 12 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Better comment?

File: src/transformers/modeling_tf_utils.py (hunk 11 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Updated!

File: src/transformers/modeling_tf_utils.py (hunk 9 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Updated!

File: src/transformers/modeling_tf_utils.py (hunk 10 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Updated!

File: src/transformers/modeling_tf_utils.py (hunk 7 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:316]: Like this?

**Comment by sgugger** [comment]
It's a lot clearer, thanks. There are still unaddressed comments however, and I can't comment on line 259 but it should be removed now (since the dict is create two lines below).

**Comment by jplu** [comment]
What is missing now?

**Comment by sgugger** [comment]
There is Lysandre's comments at line 283 and mine about the loop line 277. Like I said in my previous comments, doing the two functions in one is great, I just don't get the added complexity of the new `model_layers_name_value` variable when we could stick to the previous loop in the function `load_tf_weights` while adding the behavior of `detect_tf_missing_unexpected_layers`.

The comments are a great addition, thanks a lot for adding those!

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:267]: I'm really sorry @LysandreJik I completely forgot to answer :( And yes this is exactly for that :)

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Yes, this is expected. I need to have the name of all the layers from the checkpoint on one side, and from the instantiated model on the other side, in order to properly compute the missing and unexpected layers.

File: src/transformers/modeling_tf_utils.py (hunk 8 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:278]: You are right, this variable was not really necessary and the previous loop could be kept as it was. I changed it.

**Comment by jplu** [comment]
I have addressed the Lysandre's comment at line 283 and your for the loop at line 277. Do you see anything else?

**Review by patrickvonplaten — APPROVED** [review]
It's a lot cleaner now and good to merge for me! 

It's a nit, but there are still the "instanciated" instead of "instantiated" comments.

File: src/transformers/modeling_tf_utils.py (hunk 9 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_tf_utils.py]: seems to be still there...

**Comment by jplu** [comment]
The typos should be fixed now. Sorry for that.

**Comment by sgugger** [comment]
Good to merge for me too!

**Review by LysandreJik — APPROVED** [review]
Thank you for iterating on this, this looks good to me.
